### PR TITLE
[#7110][docdb] Add Clang thread safety annotations to YQLPartitionsVTable

### LINKED ISSUE #7110: [docdb] Add TSAN thread annotations to YQLPartitionsVTable
As a follow up to #7109 we realized we probably should leverage thread annotations to make sure we don't break this in the future. cc @nspiegelberg

### PR DESCRIPTION
**Pull request by bozhiyou**
Fixes #7110.

### CODE REVIEW

**Comment by bozhiyou** [comment]
@bmatican Would you please review this pull request? Also #8357. They are my first attempts so I want some feedback before further helping with other issues. Thanks!

File: src/yb/master/yql_partitions_vtable.h [hunk 1]
```diff
@@ -32,9 +33,9 @@ class YQLPartitionsVTable : public YQLVirtualTable {
   Schema CreateSchema() const;
 
   mutable boost::shared_mutex mutex_;
-  mutable std::shared_ptr<QLRowBlock> cache_;
-  mutable int cached_tablets_version_ = -1;
-  mutable int cached_tablet_locations_version_ = -1;
+  mutable std::shared_ptr<QLRowBlock> cache_ GUARDED_BY(mutex_);
+  mutable int cached_tablets_version_ GUARDED_BY(mutex_) = -1;
```
Comment (jmeehan16) [inline src/yb/master/yql_partitions_vtable.h:37]: This annotation and the one on `cached_tablet_locations_version_` are both good additions, but unfortunately lead to some compiler errors under certain build circumstances. This is caused by a failure of some compilers to recognize `std::shared_lock` as fulfilling the mutex-hold requirement.

Fortunately, there's an easy fix for this. Yugabyte has a custom `SharedLock` implementation (src/yb/util/shared_lock.h) with a workaround for the compiler issue. If you replace all instances of `std::shared_lock` in `yql_partitions_vtable.cc` with `SharedLock`, that should resolve the issue.

**Review by jmeehan16 — CHANGES_REQUESTED** [review]
Overall the changes are good! There are some older uses of `std::shared_lock` in `yql_partitions_vtable.cc` that can cause compile errors with these additions. However, there's an easy fix for this (see in-line comment). Once that change is made, this should be good to go!

**Comment by jmeehan16** [comment]
One other small request: would you mind adding your Git username to the [CONTRIBUTORS.md](https://github.com/yugabyte/yugabyte-db/blob/master/CONTRIBUTORS.md) file? This helps us keep track of all contributors to Yugabyte-DB.  Thanks!

File: src/yb/master/yql_partitions_vtable.h (hunk 1 above)
Comment (bozhiyou) [reply src/yb/master/yql_partitions_vtable.h:37]: Fixed. Thanks for the instructions!

**Comment by bozhiyou** [comment]
Hey @jmeehan16, I've appended my Github username to CONTRIBUTORS.md in another pull request #8495 w.r.t. my earlier merged changes. Could you please merge that one?

**Review by jmeehan16 — APPROVED** [review]
Looks good! Everything compiles fine now and passes our test suite. Thanks!

**Pull request merged**